src/main/java/gregtech/api/capability/IMultipleTankHandler.java

```
@@ -1,21 +1,210 @@
+            FluidStack tankStack = fluidTanks.get(i).getFluid();
+            if (fluidStack == tankStack || tankStack != null && tankStack.isFluidEqual(fluidStack)) {
+                return i;
+            }
+        }
+        return -1;
+    }
+
+    @Override
+    default Iterator<MultiFluidTankEntry> iterator() {
+        return getFluidTanks().iterator();
+    }
+
+    /**
+     * Entry of multi fluid tanks. Retains reference to original {@link IMultipleTankHandler} for accessing
+     * information such as {@link IMultipleTankHandler#allowSameFluidFill()}.
+     */
+    final class MultiFluidTankEntry implements IFluidTank {
+
+        private final IMultipleTankHandler tank;
+        private final IFluidTank delegate;
+
+        public MultiFluidTankEntry(@Nonnull IMultipleTankHandler tank, @Nonnull IFluidTank delegate) {
+            this.tank = tank;
+            this.delegate = delegate;
+        }
+
+        @Nonnull
+        public IMultipleTankHandler getTank() {
+            return tank;
+        }
+
+        @Nonnull
+        public IFluidTank getDelegate() {
+            return delegate;
+        }
+
+        public boolean allowSameFluidFill() {
+            return tank.allowSameFluidFill();
+        }
+
+        @Nonnull
+        public IFluidTankProperties[] getTankProperties() {
+            return delegate instanceof IFluidHandler ?
+                    ((IFluidHandler) delegate).getTankProperties() :
+                    new IFluidTankProperties[]{new FallbackTankProperty()};
+        }
+
+        @SuppressWarnings("unchecked")
+        public NBTTagCompound trySerialize() {
+            if (delegate instanceof FluidTank) {
+                return ((FluidTank) delegate).writeToNBT(new NBTTagCompound());
+            } else if (delegate instanceof INBTSerializable) {
+                try {
+                    return ((INBTSerializable<NBTTagCompound>) delegate).serializeNBT();
+                } catch (ClassCastException ignored) {}
```

Is this try-catch block needed?

---

`INBTSerializable` implementations can return other types of `NBTBase`, so the try-catch accounts for that.

---

How frequently will this catch an exception? If it's infrequent, should a log with the error be made here?

---

I think ignoring the exception is better than logging things, since logs can get quite spammy. Also I don't see how I should write the log with actually helpful information that isn't just a bunch of stacktrace and toString of the delegate instance.

Fluid tanks using different NBT type just means it's incompatible with save/load functionality of FluidTankList. If we're logging save/load failure for incompatible `INBTSerializable`s, we might as well log the cases where the delegate doesn't implement `INBTSerializable` from the start.